docs/design/coding-conventions.md

```diff
@@ -0,0 +1,413 @@
+
+Macros should in general use `SCREAMING_SNAKE_CASE`.
+It is important to prefix all macros (e.g., with `SLANG_`) to avoid collisions, since `namespace`s don't affect macros).
+
+### Prefixes
+
+Prefixes based on types (e.g., `p` for pointers) should never be used.
+
+Global variables should have a `g` prefix, e.g. `gCounter`.
+Non-`const` `static` class members can have an `s` prefix if that suits your fancy.
+Of course, both of these should be avoided, so this shouldn't come up often.
+
+Constant data (in the sense of `static const`) should have a `k` prefix.
+
+In `class` types where "information hiding" is relevant/important, an `m` prefix on members is allowed (but not required). Exmaple: `mBegin`.
+In all other cases, member variables/fields should not be prefixed.
+
+No prefixing convention is used to differentiate the public interface of an C++ feature/class from its implementation.
+
+Public C APIs will prefix all symbol names while following the casing convention (e.g. `SlangModule`, `slangLoadModule`, etc.).
+
+### Enums
+
+C-style `enum` should use the following convention:
+
+```c++
+enum Color
+{
+    kColor_Red,
+    kColor_Green,
+    kColor_Blue,
+
+    kColorCount,
+};
+```
+
+When using `enum class`, drop the `k` prefix, but retain the `UpperCamelCase` tag names:
+
+```c++
+enum class Color
+{
+    Red,
+    Green,
+    Blue,
+
+    Count,
+};
+```
+
+When defining a set of flags, separate the type definition from the `enum`:
+
+```c++
+typedef unsigned int Axes;
+enum
+{
+    kAxes_None = 0,
+
+    kAxis_X = 1 << 0,
+    kAxis_Y = 1 << 1,
+    kAxis_Z = 1 << 2,
+
+    kAxes_All = kAxis_X | kAxis_Y | kAxis_Z,
+};
+```
+
+In public APIs, `enum` cases should use `SCREAMING_SNAKE_CASE`:
+
+```c++
+typedef unsigned int SlangAxes;
+enum
+{
+    SLANG_AXES_NONE = 0,
+
+    SLANG_AXIS_X = 1 << 0,
+    SLANG_AXIS_Y = 1 << 1,
+    SLANG_AXIS_Z = 1 << 2,
+
+    SLANG_AXES_ALL = SLANG_AXIS_X | SLANG_AXIS_Y | SLANG_AXIS_Z,
+};
+```
+
+### General
+
+Names should default to the English language and US spellings, to match the dominant conventiosn of contemporary open-source projects.
+
+Function names should either be named with action verbs (`get`, `set`, `create`, `emit`, `parse`, etc.) or read as questions (`isEnabled`, `shouldEmit`, etc.).
+
+Whenever possible, compiler concepts should be named using the most widely-understood term available: e.g., we use `Token` over `Lexeme`, and `Lexer` over `Scanner` simply because they appear to be the more common names.
+
+Avoid abbreviations and initialisms unless they are already widely established across the codebase; a longer name may be cumbersome to write in the moment, but the code will probably be read many more times than it is written, so clarity should be preferred.
+An important exception to this is common compiler soncepts or techqniues which may have laboriously long names: e.g., Static Single Assignment (SSA), Sparse Conditional Copy Propagation (SCCP), etc.
```

Comment:
soncepts

Reply:
This also doesn't describe how such cases should be handled. In most slang code they are all caps. There have been studies that show using camel case is more readable, and is something that is also supported as standard for other languages. Personally from originally being in the all caps camp, I have for several years come down on the camel case style.

Reply:
My strong preference is to leave acronyms/initialisms in all caps in identifiers, because this matches the dominant convention in written English where such terms are *never* written using mixed case. It reads really weirdly to me to have an `HttpThing` when the term is "HTTP" not "Http" in written English.

I admit that the all-caps convention gets messy when it would lead to `XMLHTTPRequest` instead of `XmlHttpRequest`, but adjoined initialisms should hopefully be rare.

Any chance of getting a link to the studies? I'd be interested in whether they are speaking to exactly this issue (`OpenGLThing` vs `OpenGlThing`) or to the simpler issue of `camelCase` vs., e.g., `SCREAMING_SNAKE_CASE`.

Reply:
I'll see if I can find the study - I think Microsoft did it. That said their convention is actually slightly more convoluted

 https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions#Anchor_0

It has 2 letter acronyms staying capitalized. I personally don't think this is necessary or desirable.

I understand your argument and was in the same camp, until I worked for a company which had the camel case style, and I fairly quickly came around to it's superiority.